x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/ElasticInferenceServiceHttpClientManager.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.http;
+
+import org.apache.http.config.Registry;
+import org.apache.http.config.RegistryBuilder;
+import org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager;
+import org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor;
+import org.apache.http.impl.nio.reactor.IOReactorConfig;
+import org.apache.http.nio.conn.NoopIOSessionStrategy;
+import org.apache.http.nio.conn.SchemeIOSessionStrategy;
+import org.apache.http.nio.conn.ssl.SSLIOSessionStrategy;
+import org.apache.http.nio.reactor.ConnectingIOReactor;
+import org.apache.http.nio.reactor.IOReactorException;
+import org.apache.http.pool.PoolStats;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.XPackPlugin;
+import org.elasticsearch.xpack.core.ssl.SSLService;
+import org.elasticsearch.xpack.inference.logging.ThrottlerManager;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.inference.services.elastic.ElasticInferenceServiceSettings.EIS_SSL_ENABLED;
+import static org.elasticsearch.xpack.inference.services.elastic.ElasticInferenceServiceSettings.SSL_CONFIGURATION_PREFIX;
+import static org.elasticsearch.xpack.inference.services.elastic.ElasticInferenceServiceSettings.SSL_CONFIGURATION_SETTINGS;
+
+public class ElasticInferenceServiceHttpClientManager implements Closeable {
+    private static final Logger logger = LogManager.getLogger(HttpClientManager.class);
+    /**
+     * The maximum number of total connections the connection pool can lease to all routes.
+     * From googling around the connection pools maxTotal value should be close to the number of available threads.
+     *
+     * https://stackoverflow.com/questions/30989637/how-to-decide-optimal-settings-for-setmaxtotal-and-setdefaultmaxperroute
+     */
+    public static final Setting<Integer> MAX_TOTAL_CONNECTIONS = Setting.intSetting(
```

Comment:
Ideally we would have a single connection manager and a single set of `...http.max_total_connections` settings. I'm assuming it is not possible to use a single connection manager due to the differing SSL configurations. 

> From googling around the connection pools maxTotal value should be close to the number of available threads. 

This comment applies to blocking IO and our code is non blocking in which case it is safe to have many more connections than threads. Theoretically the limit on the number of connections is the number of ports available but we want to have a limit much lower than that. 

> I definitely see some value in being able to control these settings separately for a power user, but this is anyway not exposed in serverless at least...

Can the operator change these settings in serverless? If so that is useful. In hosted cloud the user can change these values. 

I think it is a mistake that we do not have a max value in these settings to prevent it being change to an very large number. Let's add a max value to both groups of settings (for this and the `HttpClientManager` settings) so that the combined totals cannot be an outrageously large number. 

@jonathan-buttner Is the expert, please wait for his input

Reply:
@timgrein @davidkyle, with a new [commit](https://github.com/elastic/elasticsearch/pull/116423/commits/5e876682b9d0938b37a3b600f35e032e147f850d) I simplified the HTTP client management. We will just keep `HTTPClientManager` that now supports TLS configs. However, we can still define different configurations for the Elastic Inference Service if we need. Could you please have a look again?

Reply:
Made a [separate review](https://github.com/elastic/elasticsearch/pull/116423#pullrequestreview-2469465280)